From 788eb0ca2363fb6417aa5d5402368568bce254e4 Mon Sep 17 00:00:00 2001
From: Igor Grinberg <igrinberg@habana.ai>
Date: Thu, 04 Oct 2018 23:19:51 +0300
Subject: [PATCH] flash_burn: fix the cksum leftover

The printf code fails to convert the SHA256 digest string to an integer
and returns with an error printed to stderr. The stderr is not noted by
armcp as popen() call only brings the stdout as a returned stream.
It also effectively sets the CKSUM variable value to 0.

Later on, while testing the calculated SHA256 value with the provided
one, the if statement uses the integer comparison and is also failing
as it recieves arguments that cannot be cast implicitly to integer.
While failing, it writes an error message to stderr, but gets unnoticed
by armcp for the same reason as above. The result of the if statement
is 'false' and therefore the if command list is not executed.
Effectively meaning that the digests are match.

The result of the above behavior is effectively disabled SHA256 check.
The cksum code used to check the cksum digest as an integer value.
Once we switch to SHA256 digest, we can no longer check it as a single
integer value (256 bit is 8 integer values).

Switch to string comparison instead.

Signed-off-by: Igor Grinberg <igrinberg@habana.ai>
---

diff --git a/board/habanalabs/common/rootfs_overlay/usr/sbin/flash_burn b/board/habanalabs/common/rootfs_overlay/usr/sbin/flash_burn
index 31ce393..7e7a0ab 100755
--- a/board/habanalabs/common/rootfs_overlay/usr/sbin/flash_burn
+++ b/board/habanalabs/common/rootfs_overlay/usr/sbin/flash_burn
@@ -11,7 +11,7 @@
 	while [ $# -gt 0 ]; do
 		case ${1} in
 			-c)
-				CKSUM=`printf "%d" ${2}`
+				CKSUM="${2}"
 				shift
 				shift
 			;;
@@ -48,7 +48,7 @@
 
 if [ -n "${CKSUM}" ]; then
 	SHA256=`sha256sum ${INFILE} | cut -d' ' -f1`
-	if [ "${CKSUM}" -ne "${SHA256}" ]; then
+	if [ "${CKSUM}" != "${SHA256}" ]; then
 		echo "Mismatch with provided checksum [expected ${CKSUM} while calculated ${SHA256}]"
 		exit 2
 	fi
